File: llm_core/providers/anthropic.py
import os
from collections.abc import AsyncGenerator, Sequence
from typing import Any

from anthropic import AsyncAnthropic
from llm_core.interfaces import LLMProvider

class AnthropicProvider(LLMProvider):
    """Anthropic Claude provider."""

    def __init__(self, api_key: str | None = None, model: str = "claude-3-5-sonnet-20241022"):
        self.client = AsyncAnthropic(api_key=api_key or os.environ.get("ANTHROPIC_API_KEY"))
        self.model = model

    async def generate(
        self,
        prompt: str,
        system: str | None = None,
        history: Sequence[dict[str, str]] | None = None,
        **kwargs: Any
    ) -> str:
        messages = self._build_messages(prompt, history)

        response = await self.client.messages.create(
            model=self.model,
            max_tokens=kwargs.get("max_tokens", 4096),
            system=system or "",
            messages=messages,
        )
        # response.content is a list of blocks, usually text.
        text_blocks = [block.text for block in response.content if block.type == 'text']
        return "".join(text_blocks)

    async def stream(
        self,
        prompt: str,
        system: str | None = None,
        history: Sequence[dict[str, str]] | None = None,
        **kwargs: Any
    ) -> AsyncGenerator[str, None]:
        messages = self._build_messages(prompt, history)

        async with self.client.messages.stream(
            model=self.model,
            max_tokens=kwargs.get("max_tokens", 4096),
            system=system or "",
            messages=messages,
        ) as stream:
             async for text in stream.text_stream:
                 yield text

    def _build_messages(self, prompt: str, history: Sequence[dict[str, str]] | None) -> list[dict[str, any]]:
        msgs = []
        if history:
            for h in history:
                # Anthropic message format: role (user/assistant) and content
                # Mapping model -> assistant if necessary, though 'model' is common in some schemas
                role = h.get("role", "user")
                if role == "model":
                    role = "assistant"
                msgs.append({"role": role, "content": h.get("content", "")})

        msgs.append({"role": "user", "content": prompt})
        return msgs

File: apps/api/src/llm-core/providers/anthropic.ts
import Anthropic from '@anthropic-ai/sdk';
import type { LLMProvider, ChatMessage } from '../interfaces';

export class AnthropicProvider implements LLMProvider {
  private client: Anthropic;
  private model: string;

  constructor(apiKey: string, model: string = 'claude-3-5-sonnet-20241022') {
    this.client = new Anthropic({ apiKey });
    this.model = model;
  }

  async generate(
    prompt: string,
    options?: { system?: string; history?: ChatMessage[] }
  ): Promise<string> {
    const messages = this.buildMessages(prompt, options?.history);
    
    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: 4096,
      system: options?.system,
      messages,
    });

    const textBlock = response.content.find(block => block.type === 'text');
    return textBlock ? 'text' in textBlock ? textBlock.text : '' : '';
  }

  async *stream(
    prompt: string,
    options?: { system?: string; history?: ChatMessage[] }
  ): AsyncGenerator<string> {
    const messages = this.buildMessages(prompt, options?.history);
    
    const stream = await this.client.messages.stream({
      model: this.model,
      max_tokens: 4096,
      system: options?.system,
      messages,
    });

    for await (const event of stream) {
      if (event.type === 'content_block_delta' && event.delta.type === 'text_delta') {
        yield event.delta.text;
      }
    }
  }

  private buildMessages(
    prompt: string,
    history?: ChatMessage[]
  ): Array<{ role: 'user' | 'assistant'; content: string }> {
    const messages: Array<{ role: 'user' | 'assistant'; content: string }> = [];
    
    if (history) {
      for (const msg of history) {
        if (msg.role === 'user' || msg.role === 'assistant') {
          messages.push({ role: msg.role, content: msg.content });
        }
      }
    }
    
    messages.push({ role: 'user', content: prompt });
    return messages;
  }
}
